File: Python/Classic Puzzles/Easy/mars-lander-episode-1.py
# Read the number of landing surfaces
surface_n = int(input())

# Read the coordinates of each landing surface
for i in range(surface_n):
    land_x, land_y = [int(j) for j in input().split()]  # (x, y) coordinates of the landing zone

# Main game loop
while True:
    # Read the current state of the ship
    x, y, h_speed, v_speed, fuel, rotate, power = [int(i) for i in input().split()]

    # Adjust engine power based on vertical speed and altitude
    if v_speed > 20 or y - land_y < 500:
        pow = 4  # Full power to slow descent
    else:
        pow = 3  # Reduced power for stable landing

    # Send commands: 0° rotation and the calculated power
    print(f"0 {pow}")

File: JavaScript/Classic Puzzles/Easy/mars-lander-episode-1.js
const surfaceN = parseInt(readline());

let landX1, landY1, landX2, landY2;

for (let i = 0; i < surfaceN; i++) {
    const inputs = readline().split(' ');
    const landX = parseInt(inputs[0]);
    const landY = parseInt(inputs[1]);

    if (i > 0 && landY === landY1) {
        // Found the flat landing zone
        landX2 = landX;
        landY2 = landY;
    }

    landX1 = landX;
    landY1 = landY;
}

while (true) {
    const inputs = readline().split(' ');

    const X = parseInt(inputs[0]);
    const Y = parseInt(inputs[1]);
    const hSpeed = parseInt(inputs[2]);
    const vSpeed = parseInt(inputs[3]);
    const fuel = parseInt(inputs[4]);
    const rotate = parseInt(inputs[5]);
    const currentPow = parseInt(inputs[6]);

    let power;

    // Basic vertical-speed control
    if (vSpeed < -40) {
        power = 4;
    } else if (vSpeed < -20) {
        power = 3;
    } else {
        power = 2;
    }

    console.log("0 " + power);
}
